### PR TITLE
Add yaml schema validation

### PR DESCRIPTION
**Pull request by martamatos**
This PR addresses issue #8 

However, there tests that should fail but do not, basically if there is a key that is not part of the schema i would like the validation to fail, since those keys are most likely typos, but the validation passes. 
Ideas on what to do here?

### CODE REVIEW

File: tests/test_yaml2sbml/ode_input_empty_section.yaml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    - stateId: S2
+      rightHandSide: MM(S1, Vmm, Km) - hill(S2, Shalve, Vh, h)
+      initialValue: 1
+
+    - stateId: S3
+      rightHandSide: hill(S2, Shalve, Vh, h) - k4*S3 + log10(c1)
+      initialValue: 1
+
+
+observables:
```
Comment (jvanhoefer) [inline tests/test_yaml2sbml/ode_input_empty_section.yaml:95]: Why does it have an empty observable section?

Comment (jvanhoefer) [reply tests/test_yaml2sbml/ode_input_empty_section.yaml:95]: Ah, got it :D

File: tests/test_yaml_validation.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        file_in = os.path.join(self.test_folder, 'ode_input1.yaml')
+        validate_yaml(file_in)
+
+    #Should throw an error because of extra fields in conditions
+    def test_validate_yaml_valid_2(self):
+        file_in = os.path.join(self.test_folder, 'ode_input2.yaml')
+        validate_yaml(file_in)
+
+    # Should throw an error because of typos
+    def test_validate_yaml_typos(self):
```
Comment (jvanhoefer) [inline tests/test_yaml_validation.py:25]: shouldn't a `with self.assertRaises(...):` be here?

File: tests/test_yaml_validation.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    def setUp(self):
+        this_dir, _ = os.path.split(__file__)
+        self.test_folder = os.path.join(this_dir, 'test_yaml2sbml')
+
+    def test_validate_yaml_valid_1(self):
+        file_in = os.path.join(self.test_folder, 'ode_input1.yaml')
+        validate_yaml(file_in)
+
+    #Should throw an error because of extra fields in conditions
+    def test_validate_yaml_valid_2(self):
```
Comment (jvanhoefer) [inline tests/test_yaml_validation.py:20]: Shouldn't be a `with self.assertRaises(...):` be here, too?

File: yaml2sbml/yaml_schema.yaml [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        nominalValue:
+          type: number
+          description: the parameter's value. This is optional for SBML/PEtab generation, but will be needed
+            for model simulation.
+        parameterName:
+          type: string
+          description: the parameter name. This is optional.
+        parameterScale:
+          type: string
+          description: whether the parameter is in log (log10) or linear (lin) scale. This is optional.
```
Comment (jvanhoefer) [inline yaml2sbml/yaml_schema.yaml:36]:
```suggestion
          description: whether the parameter is in `log`, `log10` or `lin` (i.e. linear)  scale. This is optional.
```

Comment (jvanhoefer) [reply yaml2sbml/yaml_schema.yaml:36]: (Since `log` (for base e) and `log10` (for base 10) are both valid parameter scales in PEtab)

**Review by jvanhoefer — APPROVED** [review]
Looks nice, I have only few questions/comments :)

File: tests/test_yaml2sbml/ode_input_empty_section.yaml (hunk 1 above)
Comment (martamatos) [reply tests/test_yaml2sbml/ode_input_empty_section.yaml:95]: Failing test xD

File: tests/test_yaml_validation.py (hunk 3 above)
Comment (martamatos) [reply tests/test_yaml_validation.py:20]: Yes, but it doesn't fail :/

File: tests/test_yaml_validation.py (hunk 2 above)
Comment (martamatos) [reply tests/test_yaml_validation.py:25]: Same as above, it should, but the function is running just fine :/

**Comment by jvanhoefer** [comment]
I will have a look at the non-failing tests today :) ...

File: tests/test_yaml_validation.py (hunk 3 above)
Comment (jvanhoefer) [reply tests/test_yaml_validation.py:20]: To be honest, I do not see, why the input yaml should result in an error... @martamatos can you help me here? :)

**Comment by jvanhoefer** [comment]
In case, you (=@martamatos) agree, that `test_validate_yaml_valid_2 ` should be fine, then I would suggest to take the comment in the line above the test definition and merge

**Comment by jvanhoefer** [comment]
I know, the code before was shorter + cleaner, but checking every yaml before it gets processed led to circular dependencies...

**Comment by martamatos** [comment]
WoW! Great job you did! Thank you! :D

Regarding `test_validate_yaml_valid_2`, it's a tricky thing, sorry for not explaining it properly :/ 
I don't think i have even defined the `yaml_schema` properly for that.
My issue is that  for each condition we can re-define parameter values, however those parameters can have pretty much any name, so i don't really know how to include that on the `yaml_schema`. 
Does it make sense?
Do you have any idea how to do this?

**Comment by jvanhoefer** [comment]
No, good point. I don't know, if this can be accessed via the `yaml_schema`. But it should currently accessed via the PEtab linter in `validate_petab_tables`. This is called by `yaml2sbml.yaml2petab` after the PEtab tables were created. So while `yaml_validation` does not catch this, `yaml2PEtab` will throw an error/warning if this happens :)

**Comment by jvanhoefer** [comment]
(to some degree this is similar to the quantities in the ODEs right hand side. They also have to be parameters/states/math-constants)...

**Comment by martamatos** [comment]
Yeah... I would say this is ready for merging, i'll poke around a bit more as soon as i find the time, but there is a good chance that it is fine as is. So if you agree give me a thumbs up and i'll merge this xD

**Comment by jvanhoefer** [comment]
I just added one small thing: In case the right hand side is constant and given as a number (e.g. `d/dt x = 5`), the validator would throw an error. Fixed this.

**Comment by jvanhoefer** [comment]
Now I am fine to merge :) 👍

**Comment by martamatos** [comment]
Nice! Thank you! :D

**Pull request merged**